File: serve/crawl/macd.go
package crawl

func (p *Tdatas) Macd() {
	macd(p, 12, 26, 9)
}

func macd(td *Tdatas, short, long, m int) {
	c := len(td.Data)
	if c < 1 {
		return
	}
	td.Data[0].emas = 0
	td.Data[0].emal = 0
	td.Data[0].DIFF = 0
	td.Data[0].DEA = 0
	td.Data[0].MACD = 0
	s1, s2 := short-1, short+1
	l1, l2 := long-1, long+1
	m1, m2 := m-1, m+1
	if c > 1 {
		td.Data[1].emas = (td.Data[0].Close*s1 + td.Data[1].Close*200) / s2
		td.Data[1].emal = (td.Data[0].Close*l1 + td.Data[1].Close*200) / l2
		td.Data[1].DIFF = td.Data[1].emas - td.Data[1].emal
		td.Data[1].DEA = (td.Data[0].DEA*m1 + td.Data[1].DIFF*2) / m2
		td.Data[1].MACD = 2 * (td.Data[1].DIFF - td.Data[1].DEA)
	}
	for i := 2; i < c; i++ {
		td.Data[i].emas = (td.Data[i-1].emas*s1 + td.Data[i].Close*200) / s2
		td.Data[i].emal = (td.Data[i-1].emal*l1 + td.Data[i].Close*200) / l2
		td.Data[i].DIFF = td.Data[i].emas - td.Data[i].emal
		td.Data[i].DEA = (td.Data[i-1].DEA*m1 + td.Data[i].DIFF*2) / m2
		td.Data[i].MACD = 2 * (td.Data[i].DIFF - td.Data[i].DEA)
	}
}

File: app/stock/macd.js
import d3 from 'd3';
import KLine from './kline';

const bColor = (d) => (d.MACD > 0) ? '#f00' : '#080';

const formatValue = d3.format(',.3f');

function fmtMacd(d) {
  return formatValue(d / 1000);
}

function dragmove() {
  d3.select(this)
    .attr('transform', `translate(0, ${d3.event.y})`);
}

class KLineMacd {
  constructor(root) {
    this.root = root;
    this.options = KLine.extend({}, this.root.options.macd);
  }

  init() {
    const height = this.height = 50;
    const rsvg = this.root._ui.svg;
    const w = +d3.select(rsvg.node().parentNode).attr('width');
    this.w = w;
    if (this.options.container) {
      const margin = this.root.options.margin;
      const container = d3.select(this.options.container);
      this.svg = container.append('svg')
        .attr('width', w)
        .attr('height', height)
        .append('g')
        .attr('transform', `translate(${margin.left},0)`);
    } else {
      const drag = d3.behavior.drag().on('drag', dragmove);
      this.svg = rsvg.append('g')
        .call(drag);
    }
  }

  hide() {
    this.svg
      .transition()
      .style('display', 'none');
  }

  show() {
    this.svg
      .transition()
      .style('display', '');
  }

  updateAxis(data) {
    if (this.root.param('nmacd')) {
      this.hide();
      return;
    }

    const svg = this.svg;
    let axis = svg.select('#macd_y_axis');
    if (axis.empty()) {
      axis = svg.append('g')
        .attr('class', 'y axis')
        .attr('id', 'macd_y_axis');
      this.y = d3.scale.linear()
        .range([this.height, 0]);
      this.yAxis = d3.svg.axis()
        .scale(this.y)
        .orient('left')
        .ticks(4)
        .tickSize(-this.w)
        .tickFormat(fmtMacd);
    }
    const min = d3.min(data, d => Math.min(d.DIFF, d.DEA, d.MACD));
    const max = d3.max(data, d => Math.max(d.DIFF, d.DEA, d.MACD));
    this.y.domain([min, max]);
    axis.call(this.yAxis);
    this.show();
  }

  update(data) {
    if (this.root.param('nmacd')) {
      this.hide();
      return;
    }
    this.show();

    const x = this.root._ui.x;
    const y = this.y;
    const candleWidth = this.root.options.candle.width;
    const svg = this.svg;

    const rect = svg.selectAll('rect.macd')
      .data(data);

    rect
      .exit()
      .transition()
      .remove();
    rect
      .enter()
      .append('rect')
      .attr('class', 'macd')
      .attr('width', candleWidth);
    rect
      .transition()
      .attr('x', (d, i) => x(i) - candleWidth / 2)
      .attr('y', (d) => y(Math.max(d.MACD, 0)))
      .attr('height', (d) => Math.abs(y(0) - y(d.MACD)))
      .attr('stroke', bColor)
      .attr('fill', bColor);

    let ldiff = svg.select('path#diff');
    if (ldiff.empty()) {
      ldiff = svg.append('path')
        .attr('id', 'diff')
        .style('fill', 'none')
        .style('stroke', 'silver')
        .style('stroke-width', '1');
    }
    ldiff.data([data]);

    let line = d3.svg.line()
      .x((d, i) => x(i))
      .y((d) => y(d.DIFF));

    ldiff.attr('d', line);

    let ldea = svg.select('path#dea');
    if (ldea.empty()) {
      ldea = svg.append('path')
        .attr('id', 'dea')
        .style('fill', 'none')
        .style('stroke', 'gold')
        .style('stroke-width', '1');
    }
    ldea.data([data]);

    line = d3.svg.line()
      .x((d, i) => x(i))
      .y((d) => y(d.DEA));

    ldea.attr('d', line);
  }
}

KLine.register_plugin('macd', KLineMacd);
